mautic/mautic
# Segment fail notifications

Note: The code changes were made by Anton, I just cherry-picked them so all the glory goes to him.

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | /
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
We want to know what was the last time we have rebuilt each segment and raise a warning if it is longer than expected. 

Acceptance criteria
- [x] Add a new column to the lead_lists table called "last_built_date"
- [x] Update timestamp every time the particular segment is being rebuilt for all contacts (not just for some). In the command for the community (when called without min, max and contact ID arguments) and in the finalize job for the cloud.
- [x] Check if last_built_date is older than 30 (configurable) hours then add a red exclamation mark next to the segment name in the segment list table with tooltip "This segment hasn't been rebuilt for X hours".

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
This is not a bug so there is nothing to reproduce.

#### Steps to test this PR:
2. Create a new segment.
3. Go to the `lead_lists` table and make sure that `last_built_date` column has datetime value. Remember that value.
4. Add some contacts to your segment using a segment filter. E.g. you can create a email based segment filter to add all contacts that have email defined (not empty).
5. Make sure that `last_built_date` value has been updated.
6. Go to the Campaign Studio -> Configuration -> Segment Settings -> Show warning if segment hasn't been rebuilt for X hours. Change that value to 1.
7. Wait 1 hour, go the segments view page `/s/segments`. Make sure that you see a red exclamation mark next to the segment name. If you hover the cursor over it you should see the following text: 
"This segment hasn't been rebuilt for 1 hour."

#### Other areas of Mautic that may be affected by the change:
1. I can't think of any. 


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10772"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
